docs/cpp/source/notes/tensor_cuda_stream.rst

```diff
@@ -0,0 +1,173 @@
+  assert(curStream == myStream);
+
+  // get the default CUDA stream on current device (device 0)
+  // set current stream with default stream and get current stream
+  torch::cuda::CUDAStream defaultStream = torch::cuda::getDefaultCUDAStream();
+  torch::cuda::setCurrentCUDAStream(defaultStream);
+  curStream = torch::cuda::getCurrentCUDAStream();
+  assert(curStream == defaultStream);
+  assert(myStream != defaultStream)
+
+.. attention::
+  
+  Above code is running on the same device/gpu (say device 0). `setCurrentCUDAStream` works as expected and set current CUDA stream correctly
+  on device 0, this is because all the streams are on the same device. However, say above code is running on device 0 and we get `myStream`
+  from device 1, then `torch::cuda::setCurrentCUDAStream(myStream)` will set `myStream` as current stream on device 1, not device 0.
+
+
+2. Use various CUDA Guard to set CUDA stream
+
+.. code-block:: cpp
+
+   // create a vector of CUDA stream from current device (device 0)
+   std::vector<torch::cuda::CUDAStream> streams0 = {torch::cuda::getDefaultCUDAStream(), torch::cuda::getStreamFromPool()};
+   assert(streams0[0].device_index() == 0);
+   assert(streams0[1].device_index() == 0);
+   // set current stream as `streams0[0]` on `streams0[0]`'s current device (device 0)
+   torch::cuda::setCurrentCUDAStream(streams0[0]);
+
+   // create a vector of CUDA stream from device 1
+   std::vector<torch::cuda::CUDAStream> streams1;
+   {
+     // create a CUDA Device guard within the scope to guard device 1
+     torch::cuda::CUDAGuard device_guard(1);
+     // get default CUDA stream from device 1
+     streams1.push_back(torch::cuda::getDefaultCUDAStream());
+     // get a new stream from CUDA stream pool on device 1
+     streams1.push_back(torch::cuda::getStreamFromPool());
+   }
+   assert(streams1[0].device_index() == 1);
+   assert(streams1[1].device_index() == 1);
+   // set current stream as `streams1[0]` on `streams1[0]`'s current device (device 1)
+   torch::cuda::setCurrentCUDAStream(streams1[0]);
+   // current device is still 0 out of the scope of previous device guard
+   assert(torch::cuda::current_device() == 0);
+
+   // use CUDAStreamGuard to set a stream, it changes the stream on curernt devide and also passed in stream's device
+   {
+     torch::cuda::CUDAStreamGuard guard(streams1[1]);
+     // current device is 1 within this scope
+     assert(guard.current_device() == torch::Device(torch::kCUDA, 1));
+     assert(torch::cuda::current_device() == 1);
+     // current stream on device 1 is streams1[1], not streams1[0]
+     assert(torch::cuda::getCurrentCUDAStream(1) == streams1[1]);
+   }
+
+   // device and stream are now back to the status before CUDAStreamGuard is created.
+   assert(torch::cuda::current_device() == 0);
+   assert(torch::cuda::getCurrentCUDAStream(1) == streams1[0]);
+
+   // CUDA Device guard on changes the current device, but not the stream
```

Comment:
This is a little misleading. Changing the device does change the current stream, after all (to whatever stream is current on the device switched to).

Reply:
typo.....I want to say 'CUDA Device guard only changes the current device'